Install helper to setup CoCo on baremetal worker nodes

bpradipt opened the pull request:
Setup CoCo on baremetal OpenShift worker nodes using the layered image deployment feature of OSC operator.

Added few additional commits to move the scripts from hack to scripts folder

bpradipt committed:
Install helper to setup CoCo on baremetal worker nodes

Setup CoCo on baremetal OpenShift worker nodes using the layered
image deployment feature of OSC operator.

Signed-off-by: Pradipta Banerjee <pradipta.banerjee@gmail.com>
Signed-off-by: Niteesh Dubey <niteesh@us.ibm.com>
Signed-off-by: Fabiano Fidêncio <fabiano@fidencio.org>

bpradipt committed:
Move rvps-extraction scripts to the scripts folder

Also rename the folder to use smallcase.

Signed-off-by: Pradipta Banerjee <pradipta.banerjee@gmail.com>

bpradipt commented:
@apinnick I have moved the scripts from hack folder to scripts folder in this PR. PTAL

apinnick reviewed (approved):
LGTM

gkurz commented on scripts/install-helpers/baremetal-coco/.gitignore:
```diff
@@ -0,0 +1,4 @@
+id_rsa*
+*kubeconfig*
```
The `install.sh` script doesn't deal with kubeconfig or id_rsa files in any way AFAICT. Why is this needed ?

bpradipt replied:
Removed

gkurz commented on scripts/install-helpers/baremetal-coco/README.md:
```diff
@@ -0,0 +1,100 @@
+Note that currently the script only supports installing a single TEE environment.
+
+## Prerequisites
+
```
You might want to list the `oc` and `jq` commands there since they are required by `install.sh`.

gkurz commented on scripts/install-helpers/baremetal-coco/install.sh:
```diff
@@ -0,0 +1,733 @@
+        statusUpdating=$(oc get mcp "$mcp" -o=jsonpath='{.status.conditions[?(@.type=="Updating")].status}')
+        statusDegraded=$(oc get mcp "$mcp" -o=jsonpath='{.status.conditions[?(@.type=="Degraded")].status}')
+    done
+
```
```suggestion
    echo "MCP $mcp is not ready after $timeout seconds"
    return 1
```

gkurz commented on scripts/install-helpers/baremetal-coco/install.sh:
```diff
@@ -0,0 +1,733 @@
+    while [ $elapsed -lt "$timeout" ]; do
+        ready=$(oc get runtimeclass "$runtimeclass" -o jsonpath='{.metadata.name}')
+        if [ "$ready" == "$runtimeclass" ]; then
+            echo "Runtimeclass $runtimeclass is ready"
```
```suggestion
            echo "RuntimeClass $runtimeclass is ready"
```

gkurz commented on scripts/install-helpers/baremetal-coco/install.sh:
```diff
@@ -0,0 +1,733 @@
+    while [ $elapsed -lt "$timeout" ]; do
+        ip=$(oc get endpoints -n "$namespace" "$service" -o jsonpath='{.subsets[0].addresses[0].ip}')
+        if [ -n "$ip" ]; then
+            echo "Service $service IP is available"
```
Maybe include `"$ip"` in the output ?

gkurz replied:
This is still not printing `"$ip"`.

bpradipt replied:
:-(

gkurz commented on scripts/install-helpers/baremetal-coco/install.sh:
```diff
@@ -0,0 +1,733 @@
+        elapsed=$((elapsed + interval))
+    done
+
+    echo "Runtimeclass $runtimeclass is not ready after $timeout seconds"
```
```suggestion
    echo "RuntimeClass $runtimeclass is not ready after $timeout seconds"
```

gkurz commented on scripts/install-helpers/baremetal-coco/install.sh:
```diff
@@ -0,0 +1,733 @@
+
+# Function to apply the operator manifests
+function apply_operator_manifests() {
+    # Apply the manifests, error exit if any of them fail
```
What is the motivation to `exit 1` instead of `return 1` ?

gkurz commented on scripts/install-helpers/baremetal-coco/install.sh:
```diff
@@ -0,0 +1,733 @@
+    if [ -z "$PULL_SECRET_JSON" ]; then
+        echo "PULL_SECRET_JSON environment variable is not set"
+        echo "example PULL_SECRET_JSON='{\"my.registry.io\": {\"auth\": \"ABC\"}}'"
+        exit 1
```
Same question as with `apply_operator_manifests()`.

gkurz commented on scripts/install-helpers/baremetal-coco/install.sh:
```diff
@@ -0,0 +1,733 @@
+function deploy_nfd_operator() {
+    echo "Node Feature Discovery operator | starting the deployment"
+
+    pushd nfd || return 1
```
Since the `nfd` directory is a strong requirement here, I'd rather check its existence upfront, before calling this function and explicitly report when it is missing.

bpradipt replied:
please see https://github.com/openshift/sandboxed-containers-operator/pull/479#discussion_r1869867440

gkurz commented on scripts/install-helpers/baremetal-coco/install.sh:
```diff
@@ -0,0 +1,733 @@
+    oc apply -f ns.yaml || return 1
+    oc apply -f og.yaml || return 1
+    oc apply -f subs.yaml || return 1
+    popd || return 1
```
The only reason for `popd` to fail would be the removal of the current working directory at the time `pushd` was called... can this even happen ?

bpradipt replied:
This was added to keep shellcheck happy - https://www.shellcheck.net/wiki/SC2164

bpradipt replied:
The return statements for both pushd and popfd were dictated by shellcheck. I can disable it as well, but thought it's better to keep it.  Let me know if you have a strong preference to remove it :-)

gkurz replied:
I prefer to have shellcheck happy :wink:

gkurz commented on scripts/install-helpers/baremetal-coco/install.sh:
```diff
@@ -0,0 +1,733 @@
+
+    pushd nfd || return 1
+    oc apply -f intel-rules.yaml || return 1
+    popd || return 1
```
Same comments as with `deploy_nfd_operator()`.

gkurz commented on scripts/install-helpers/baremetal-coco/install.sh:
```diff
@@ -0,0 +1,733 @@
+    wait_for_deployment inteldeviceplugins-controller-manager openshift-operators || return 1
+    oc apply -f sgx_device_plugin.yaml || return 1
+    popd || return 1
+
```
Same comments as with deploy_nfd_operator().

gkurz commented on scripts/install-helpers/baremetal-coco/install.sh:
```diff
@@ -0,0 +1,733 @@
+    pushd nfd || return 1
+    oc apply -f amd-rules.yaml || return 1
+    popd || return 1
+
```
Same comments as with deploy_nfd_operator().

gkurz commented on scripts/install-helpers/baremetal-coco/install.sh:
```diff
@@ -0,0 +1,733 @@
+        ;;
+    esac
+
+    cat <<EOF >./96-kata-kernel-config-mc.yaml
```
This literal string as 3 other users. Maybe put it in a global ?

Also you might want to add this file to `.gitignore`.

gkurz commented on scripts/install-helpers/baremetal-coco/install.sh:
```diff
@@ -0,0 +1,733 @@
+    [ $result -eq 0 ] || return 1
+
+    echo "KataConfig object successfully created"
+    return 0
```
Not needed.
```suggestion
```

gkurz commented on scripts/install-helpers/baremetal-coco/install.sh:
```diff
@@ -0,0 +1,733 @@
+    esac
+
+    # Create KataConfig object
+    oc apply -f - <<EOF
```
If you do :
```suggestion
    oc apply -f - || return 1 <<EOF
```
you can...

gkurz commented on scripts/install-helpers/baremetal-coco/install.sh:
```diff
@@ -0,0 +1,733 @@
+    matchLabels:
+      $label
+EOF
+
+    result=$?
+    [ $result -eq 0 ] || return 1
```
... drop these lines
```suggestion
```

gkurz commented on scripts/install-helpers/baremetal-coco/install.sh:
```diff
@@ -0,0 +1,733 @@
+        path: $filepath
+EOF
+
+    oc apply -f 96-kata-kernel-config-mc.yaml || return 1
+
+}
```
The `|| return 1` isn't needed.

```suggestion
    oc apply -f 96-kata-kernel-config-mc.yaml
}
```

gkurz commented on scripts/install-helpers/baremetal-coco/install.sh:
```diff
@@ -0,0 +1,733 @@
+    echo "Creating kata-$tee_type RuntimeClass object"
+
+    #Create runtimeClass object
+    oc apply -f - <<EOF
```
If you do :
```suggestion
    oc apply -f - || return 1 <<EOF
```
you can...

gkurz commented on scripts/install-helpers/baremetal-coco/install.sh:
```diff
@@ -0,0 +1,733 @@
+  nodeSelector:
+     $label
+EOF
+    result=$?
+    [ $result -eq 0 ] || return 1
```
... drop these lines
```suggestion
```

gkurz commented on scripts/install-helpers/baremetal-coco/install.sh:
```diff
@@ -0,0 +1,733 @@
+    [ $result -eq 0 ] || return 1
+
+    echo "kata-$tee_type RuntimeClass object successfully created"
+    return 0
+
+}
```
Not needed.
```suggestion
}
```

gkurz commented on scripts/install-helpers/baremetal-coco/install.sh:
```diff
@@ -0,0 +1,733 @@
+    if [ -z "$TEE_TYPE" ]; then
+        echo "Error: TEE type (-t) is mandatory"
+        display_help
+        exit 1
```
What about `return 1` and let the caller decide ?

bpradipt replied:
Yeah, I'll make it consistent across the code.

gkurz commented on scripts/install-helpers/baremetal-coco/install.sh:
```diff
@@ -0,0 +1,733 @@
+    oc get deployment nfd-controller-manager -n openshift-nfd &>/dev/null
+    return_code=$?
+    if [ $return_code -eq 0 ]; then
+        pushd nfd || return 1
```
Same remark as with other invocations `pushd` and `popd`.

gkurz commented on scripts/install-helpers/baremetal-coco/install.sh:
```diff
@@ -0,0 +1,733 @@
+    echo "Uninstalling all the artifacts"
+
+    # Uninstall NFD
+    uninstall_node_feature_discovery "$TEE_TYPE" || exit 1
```
What about `return 1` and let the caller decide ?

gkurz commented on scripts/install-helpers/baremetal-coco/install.sh:
```diff
@@ -0,0 +1,733 @@
+oc apply -f osc-fg-cm.yaml || exit 1
+
+# Create Layered Image FG ConfigMap
+if [ "$TEE_TYPE" = "tdx" ]; then
+    oc apply -f layeredimage-cm-tdx.yaml || exit 1
+elif [ "$TEE_TYPE" = "snp" ]; then
+    oc apply -f layeredimage-cm-snp.yaml || exit 1
+else
+    echo "Unsupported TEE_TYPE. It must be tdx or snp" || exit 1
+fi
```
`TEE_TYPE` is already validated by `verify_params()`. You can just use a `case` command as already done a few lines above.

```suggestion
    case $TEE_TYPE in
    tdx)
        oc apply -f layeredimage-cm-tdx.yaml || exit 1
        ;;
    snp)
        oc apply -f layeredimage-cm-snp.yaml || exit 1
        ;;
    esac
```

gkurz commented on scripts/install-helpers/baremetal-coco/intel-dpo/sgx_device_plugin.yaml:
```diff
@@ -0,0 +1,14 @@
+  provisionLimit: 110
+  logLevel: 4
+  nodeSelector:
+    intel.feature.node.kubernetes.io/sgx: "true"
```
GH ui reports this file is lacking a terminating newline.

gkurz commented on scripts/install-helpers/baremetal-coco/intel-dpo/install_operator.yaml:
```diff
@@ -0,0 +1,23 @@
+  name: intel-device-plugins-operator
+  source: certified-operators
+  sourceNamespace: openshift-marketplace
+  startingCSV: intel-device-plugins-operator.v0.29.1
```
GH ui reports this file is lacking a terminating newline.

gkurz commented on scripts/install-helpers/baremetal-coco/nfd/intel-rules.yaml:
```diff
@@ -0,0 +1,64 @@
+  rules:
+    - name: "intel.gpu"
+      labels:
+        "intel.feature.node.kubernetes.io/gpu": "true"
```
This is supposed to be about CoCo, right ? Can you clarify  why we care for a GPU here ?

bpradipt replied:
Good point. I think this comes from confidential GPU. But right now it doesn't work with OSC anyway .. So I'll remove it

gkurz commented on scripts/install-helpers/baremetal-coco/nfd/ns.yaml:
```diff
@@ -0,0 +1,7 @@
+  name: openshift-nfd
+  labels:
+    name: openshift-nfd
+    openshift.io/cluster-monitoring: "true"	
```
GH ui reports this file is lacking a terminating newline.

gkurz commented on scripts/install-helpers/baremetal-coco/nfd/og.yaml:
```diff
@@ -0,0 +1,9 @@
+  namespace: openshift-nfd
+spec:
+  targetNamespaces:
+  - openshift-nfd
```
GH ui reports this file is lacking a terminating newline.

gkurz commented on scripts/install-helpers/baremetal-coco/osc_catalog.yaml:
```diff
@@ -0,0 +1,12 @@
+spec:
+  displayName: OSC Upstream Operator Catalog
+  sourceType: grpc
+  image: quay.io/openshift_sandboxed_containers/openshift-sandboxed-containers-operator-catalog:1.8.1-1
```
Current (and probably GA) build is 1.8.1-3.
```suggestion
  image: quay.io/openshift_sandboxed_containers/openshift-sandboxed-containers-operator-catalog:1.8.1-3
```

gkurz reviewed (commented):
Hi @bpradipt ! This looks pretty good to me. Posted some comments.

beraldoleal reviewed (commented):
@bpradipt while reading this script, helm keeps popping on my mind. Have you considered using helm for this job? Any reason for not using helm here?

bpradipt commented:
> @bpradipt while reading this script, helm keeps popping on my mind. Have you considered using helm for this job? Any reason for not using helm here?

No particular reason. We just happened to start with bash and that's all.

beraldoleal commented on scripts/install-helpers/baremetal-coco/subs.yaml:
```diff
@@ -0,0 +1,12 @@
+apiVersion: operators.coreos.com/v1alpha1
```
since we have `subs-ga.yaml`, for better organization, maybe rename this one to `subs-upstream.yaml` ?

beraldoleal commented:
> No particular reason. We just happened to start with bash and that's all.

Good to know.

I think we’d benefit a lot from using Helm. While this might not be the best place to start the effort, let’s keep the idea in mind and continue discussing it, as delaying could make migration harder later.

bpradipt commented:
> > No particular reason. We just happened to start with bash and that's all.
> 
> Good to know.
> 
> I think we’d benefit a lot from using Helm. While this might not be the best place to start the effort, let’s keep the idea in mind and continue discussing it, as delaying could make migration harder later.

We should also evaluate helm vs ansible approach imho.

gkurz commented on scripts/install-helpers/baremetal-coco/install.sh:
```diff
@@ -0,0 +1,728 @@
+        sleep $interval
+        elapsed=$((elapsed + interval))
+    done
+    echo "Service $service IP ($ip) is not available after $timeout seconds"
```
I think the `($ip)` should go to the success path.

bpradipt replied:
yeah my bad.. Sorry about it

bpradipt commented:
@gkurz @beraldoleal I think I have addressed all the comments. PTAL and let me know if I have missed something.
Also updated the SoB for the baremetal install commit based on original code in personal branch - https://github.com/bpradipt/coco-install

snir911 reviewed (approved):
LGTM thanks, maybe we can remove /hack completely now..

beraldoleal reviewed (approved):
LGTM

bpradipt committed:
Move configMap helpers under scripts directory

Place it under cm-helpers sub-dir

Signed-off-by: Pradipta Banerjee <pradipta.banerjee@gmail.com>

bpradipt commented:
Force-pushed to fix incorrect usage of EOF when using `oc apply`. Thanks to @niteeshkd for fixing it.

bpradipt commented:
@gkurz is it good to be merged ?

gkurz commented:
/test sandboxed-containers-operator-e2e

gkurz commented:
> @gkurz is it good to be merged ?

Yes it is @bpradipt, I just want to understand why the e2e CI check is failing in this PR and not in #487 .
